R/upsert2.R

```diff
@@ -0,0 +1,168 @@
+#'
+#' Upload a table to S3 and then load to Redshift, replacing the target value
+#' in existing rows that have the same keys, and inserting rows with new keys.
+#' New rows must match structure and column ordering of existing Redshift table.
+#'
+#' @param dat a data frame
+#' @param dbcon connection to the Redshift server
+#' @param table_name the name of the table to replace
+#' @param split_files how many files to split into. Default based on number of slices in Redshift.
+#' @param values the columns that reflect updated values
+#' @param keys the key columns on which to upsert
+#' @param bucket temporary AWS bucket; defaults to AWS_BUCKET_NAME
+#' @param region temporary AWS region; defaults to AWS_DEFAULT_REGION
+#' @param access_key AWS access key; defaults to AWS_ACCESS_KEY_ID
+#' @param secret_key AWS secret key; defaults to AWS_SECRET_ACCESS_KEY
+#' @param iam_role_arn IAM role; defaults to AWS_IAM_ROLE_ARN. Ignores access_key and secret_key.
+#' @param wlm_slots number of WLM slots for bulk load
+#' @param additional_params Additional paramseters to Redshift COPY statement
+#'
+#' @examples
+#' library(DBI)
+#'
+#' a=data.frame(a=seq(1,10000), b=seq(10000,1))
+#' n=head(a,n=5000)
+#' n$b=n$a
+#' nx=rbind(n, data.frame(a=seq(99999:104000), b=seq(104000:99999)))
+#'
+#'\dontrun{
+#' con <- dbConnect(RPostgres::Postgres(), dbname="dbname",
+#' host='my-redshift-url.amazon.com', port='5439',
+#' user='myuser', password='mypassword',sslmode='require')
+#'
+#' rs_cols_upsert_table(df=nx, dbcon=con, table_name='testTable',
+#' bucket="my-bucket", split_files=4, values=c('b'), keys=c('a'))
+#'}
+#' @seealso \url{https://docs.aws.amazon.com/redshift/latest/dg/merge-specify-a-column-list.html}
+#' @seealso \url{http://docs.aws.amazon.com/redshift/latest/dg/tutorial-configuring-workload-management.html}
+#' @export
+rs_cols_upsert_table = function(dat,
+                                dbcon,
+                                table_name,
+                                values,
+                                keys,
+                                split_files,
+                                bucket = Sys.getenv('AWS_BUCKET_NAME'),
+                                region = Sys.getenv('AWS_DEFAULT_REGION'),
+                                access_key = Sys.getenv('AWS_ACCESS_KEY_ID'),
+                                secret_key = Sys.getenv('AWS_SECRET_ACCESS_KEY'),
+                                iam_role_arn = Sys.getenv('AWS_IAM_ROLE_ARN'),
+                                wlm_slots = 1,
+                                additional_params = '') {
+  if (!inherits(dat, 'data.frame')) {
+    warning("dat must be a data.frame or inherit from data.frame")
+    return(FALSE)
+  }
+  numRows = nrow(dat)
+
+  if (numRows == 0) {
+    warning("Empty dataset provided, will not try uploading")
+    return(FALSE)
+  }
+
+  if (getOption("verbose")) {
+    message("The provided data.frame has ", numRows, " rows")
```

Comment:
I only made this comment once but applies to all usage of the verbose option.

Reply:
Either `supressMessages` or `getOption("verbose")` seems fine to me. Your call, happy to take out the `if`. Probably best to make it consistent across the entire package, although I'm not sure I'm up for updating the other `rs_*` functions right now.

Reply:
I'd prefer to make them appear by default, and allow them to be suppressed, because a novice user may have no idea what is going on, also i think it's a good default to debug in logs for batch processes, in case things go wrong